src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -4255,6 +4255,90 @@ module Queries = struct
+        consensus_state
+    in
+    let next_epoch =
+      Consensus.Proof_of_stake.Data.Consensus_state.next_epoch_data
+        consensus_state
+    in
+    let cast_ledger = function
+      | Consensus.Data.Local_state.Snapshot.Ledger_snapshot.Genesis_epoch_ledger
+          l ->
+          Ledger.Any_ledger.cast (module Ledger) l
+      | Consensus.Data.Local_state.Snapshot.Ledger_snapshot.Ledger_db l ->
+          Ledger.Any_ledger.cast (module Ledger.Db) l
+    in
+    let root_consensus_state =
+      let frontier =
+        Option.value_exn @@ Pipe_lib.Broadcast_pipe.Reader.peek
+        @@ Mina_lib.transition_frontier mina
+      in
+      let frontier_root = Transition_frontier.root frontier in
+      frontier_root |> Transition_frontier.Breadcrumb.protocol_state
+      |> Mina_state.Protocol_state.consensus_state
+    in
+    let%map staking_ledger, next_epoch_ledger =
+      match
+        (* We pretend that the block is finalized, so that we can query it in
+           advance, for redundancy.
+        *)
+        Consensus.Hooks.get_epoch_ledgers_for_finalized_frontier_block
+          ~root_consensus_state ~target_consensus_state:consensus_state
+          ~local_state:frontier_consensus_local_state
+      with
+      | `Both (staking_ledger, next_epoch_ledger) ->
+          return (cast_ledger staking_ledger, cast_ledger next_epoch_ledger)
+      | `Snarked_ledger (staking_ledger, num_parents) ->
+          (* The epoch transition was at a block between the current block and
+             the root. We find it by walking back by `num_parents` blocks.
+          *)
+          let%bind epoch_transition_state_hash =
+            let open Result.Let_syntax in
+            let rec ancestor breadcrumb i =
+              if i = 0 then return breadcrumb
+              else
+                let parent_hash =
+                  Transition_frontier.Breadcrumb.parent_hash breadcrumb
+                in
+                let%bind breadcrumb =
+                  Mina_lib.best_chain_block_by_state_hash mina parent_hash
+                in
+                ancestor breadcrumb (i - 1)
+            in
+            ancestor breadcrumb num_parents
+            >>| Transition_frontier.Breadcrumb.state_hash |> Deferred.return
+          in
+          (* When this block reaches the root of the frontier, its snarked
+             ledger will become the next epoch ledger; we simulate that here.
+          *)
+          let%map next_epoch_ledger =
+            Mina_lib.get_snarked_ledger_full mina
+              (Some epoch_transition_state_hash)
+            |> Deferred.Result.map_error ~f:Error.to_string_hum
+          in
+          ( cast_ledger staking_ledger
+          , Ledger.Any_ledger.cast (module Ledger) next_epoch_ledger )
+    in
+    assert (
+      Mina_base.Ledger_hash.equal
+        (Ledger.Any_ledger.M.merkle_root staking_ledger)
+        staking_epoch.ledger.hash ) ;
+    assert (
```

Comment:
Mmm.. curious, so with the current implementation this condition is expected to succeed?

Reply:
Initially I believe it was said that next epoch ledger hash of a block before `k`-th depth is probably some garbage which shouldn't be checked against.

Reply:
The circuit is generally the best place to look. I found [there](https://github.com/MinaProtocol/mina/blob/30962d7c32142ce11479cc885b8fd1ff4c3f9929/src/lib/consensus/proof_of_stake.ml#L2290-L2295) that we update it on the first block of the new epoch with the snarked ledger hash of the previous epoch, and then preserve it until the next time. The seed itself wanders around, but the ledger becomes fixed at each epoch transition. Notice also that `previous_blockchain_state_ledger_hash` is actually `snarked_ledger_hash`, where it's bound [further down the file](https://github.com/MinaProtocol/mina/blob/30962d7c32142ce11479cc885b8fd1ff4c3f9929/src/lib/consensus/proof_of_stake.ml#L3458).